pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -784,3 +799,51 @@ func (c *Controller) updateEDS(ep *v1.Endpoints) {
 		c.XDSUpdater.ConfigUpdate(false)
 	}
 }
+
+// InitNetworkLookup will read the mesh networks configuration from the environment
+// and initialize CIDR rangers for an efficient network lookup when needed
+func (c *Controller) InitNetworkLookup() {
+	if c.Env == nil || c.Env.MeshNetworks == nil {
+		return
+	}
+
+	c.networksRangers = map[cidranger.RangerEntry]string{}
+	for n, v := range c.Env.MeshNetworks.Networks {
+		for _, ep := range v.Endpoints {
+			if ep.GetFromCidr() != "" {
+				_, net, err := net.ParseCIDR(ep.GetFromCidr())
+				if err != nil {
+					log.Warnf("unable to parse CIDR %s for network %s", ep.GetFromCidr(), n)
+					continue
+				}
+				rangerEntry := cidranger.NewBasicRangerEntry(*net)
+				c.ranger.Insert(rangerEntry)
+				c.networksRangers[rangerEntry] = n
+			}
+			if ep.GetFromRegistry() != "" && ep.GetFromRegistry() == c.ClusterID {
+				c.network = n
+			}
+		}
+	}
+}
+
+// return the mesh network for the endpoint address. Empty string if not found.
+func (c *Controller) endpointNetwork(ea *v1.EndpointAddress) string {
+	if c.networksRangers == nil {
+		return ""
+	}
+
+	entries, err := c.ranger.ContainingNetworks(net.ParseIP(ea.IP))
+	if err != nil {
+		log.Errora(err)
+		return ""
+	}
+	if len(entries) == 0 {
+		return ""
+	}
+	if len(entries) > 1 {
+		log.Warnf("Found multiple networks CIDRs matching the endpoint IP: %s. Using the first match.", ea.IP)
```

Comment:
We don't support multiple networks per endpoint because k8s doesn't support it either. Instead, we opted to do duplicate the endpoints, one per network.